torch/nn/parallel/distributed.py

```diff
@@ -1091,23 +1115,27 @@ def join(
                 would want to set this to ``True`` for cases where the last few
                 inputs of your training job are uneven. In extreme cases, where
                 there is a large discrepancy in the number of inputs, setting
-                this to ``False`` might provide better results.
-            enable (bool): Whether to enable uneven input detection or not. Pass
-                in ``enable=False`` to disable in cases where you know that
-                inputs are even across participating processes. Default is
+                this to ``False`` might provide better results. Default is
+                ``True``.
+            enable (bool): Whether to enable uneven input detection or not.
+                Pass in ``enable=False`` to disable in cases where you know
+                that inputs are even across participating processes. Default is
                 ``True``.
             throw_on_early_termination (bool): Whether to throw an error
                 or continue training when at least one rank has exhausted
-                inputs. If ``True``, will throw upon the first rank reaching end
-                of data. If ``False``, will continue training with a smaller
-                effective world size until all ranks are joined. Note that if
-                this flag is specified, then the flag
-                ``divide_by_initial_world_size`` would be ignored. Default
-                is ``False``.
-
+                inputs. If ``True``, will throw upon the first rank reaching
+                end of data. If ``False``, will continue training with a
+                smaller effective world size until all ranks are joined. Note
+                that if this flag is specified, then the flag
+                ``divide_by_initial_world_size`` would be ignored. Default is
+                ``False``.
+            join_hooks (list[JoinHook]): A :class:`list` of ``JoinHook`` s
```

Comment:
Comments below are based on my personal understanding. Please feel free to disagree. 

It's a bit weird to install optimizer hooks here, because `nn.Module` and `optim.*` are separate concepts in PyTorch and owns separate scopes. Registering optimizer hooks here means that we let optimizer concepts leaking into `nn.Module` scope. I am not saying this is impossible, but if we plan to make such changes, we will need to start a discussion with the rest of PyTorch team and have holistic design to make sure that breaking the conceptual boundary is OK and future changes are coherent with the rest of PyTorch features. Given we only have a few weeks left, it might be a hard task to push. 

So, if this is just for test purpose, we can make this argument, `join_hook` and `JoinHook` private. 

Another option (I am more inclined to go with this one) is to we don't change DDP `join()` API at all. We still need to extract the join logic into a private `join_hook` function (let's start from private `_join_hook`), and we can build a generic `join()` context in (maybe) `https://torch/distributed/algorithms/` or a new `https://torch/distributed/utils/` folder. That generic join context can take a list of (custom) hooks to cover both `nn.Module`, `Optimizer`, or even `SyncBatchNorm`, etc.